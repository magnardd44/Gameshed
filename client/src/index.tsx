import * as React from 'react';
import { Component } from 'react-simplified';
import ReactDOM from 'react-dom';
import {
  Card,
  Alert,
  Row,
  Form,
  Column,
  Button,
  NavBar,
  Header,
  Container,
  ColumnCentre,
  SearchResult,
} from './widgets';
import axios from 'axios';
import { HashRouter, Route } from 'react-router-dom';
import { AddReview, PublishReview } from './review-components';
import { Hash } from 'crypto';
import { createHashHistory } from 'history';

axios.defaults.baseURL = 'http://localhost:3000/api/v2';

const history = createHashHistory();

class NavHeader extends Component {
  render() {
    return (
      <div>
        <NavBar brand="GameShed">
          <NavBar.Link to="/games">Games</NavBar.Link>
        </NavBar>
        <Header></Header>
      </div>
    );
  }
}

class Search extends Component {
  input = '';
  render() {
    return (
      <>
        <Container textalign={'centre'}>
          <Row>
            <ColumnCentre width={6} offset={2}>
              <Form.Input
                type={this.input}
                value={this.input}
                placeholder="Søk etter et spill"
                onChange={(event) => (this.input = event.currentTarget.value)}
              />
            </ColumnCentre>
            <ColumnCentre width={2}>
              <Button.Success
                onClick={() => {
                  this.search();
                }}
              >
                Søk
              </Button.Success>
            </ColumnCentre>
          </Row>
        </Container>
      </>
    );
  }
  search() {
    console.log('search');
    history.push('/results');
  }
}

class SearchListings extends Component {
  render() {
    return (
      <Container>
        Søkeresultater:
        <SearchResult></SearchResult>
        <SearchResult></SearchResult>
        <SearchResult></SearchResult>
      </Container>
    );
  }
}

ReactDOM.render(
  <HashRouter>
    <div>
      <NavHeader />
      <Route exact path="/" component={Search}></Route>
      <Route exact path="/results" component={SearchListings}></Route>
      <Route exact path="/addReview" component={AddReview} />
      <Route exact path="/publishReview/:id" component={PublishReview} />
    </div>
  </HashRouter>,
  document.getElementById('root')
);

/*
class Start extends Component {
  input = '';
  stdout = '';
  stderr = '';
  errCode: number | null = null;
}



  render() {
    return (
      <>
        <Card title="GameShed">
          <Row>
            <Column>
              <Form.Label>Søk:</Form.Label>
              <Form.Textarea
                value={this.input}
                onChange={(event) => (this.input = event.currentTarget.value)}
              />
            </Column>
          </Row>
          <Button.Success
            onClick={() => {
              axios
                .post<{ exitStatus: number; stdout: string; stderr: string }>('/run', {
                  language: 'js',
                  source: this.input,
                })
                .then((response) => {
                  this.errCode = response.data.exitStatus;
                  this.stdout = response.data.stdout;
                  this.stderr = response.data.stderr;
                })
                .catch((error: Error) => Alert.danger('Could not run app.js: ' + error.message));
            }}
          >
            Søk etter spill
          </Button.Success>
        </Card>
        <Card title="Standard output">{this.stdout}</Card>
        <Card title="Standard error">{this.stderr}</Card>
        <Card title="Errorcode">{this.errCode}</Card>
      </>
    );
  }
}


const root = document.getElementById('root');
if (root)
  ReactDOM.render(
    <HashRouter>
      <Alert />

      <Route exact path="/" component={Start} />
      
    </HashRouter>,
    root
  );
 
 */
