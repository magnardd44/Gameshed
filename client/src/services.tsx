import axios from 'axios';
import { sharedComponentData } from 'react-simplified';

axios.defaults.baseURL = 'http://localhost:3000/api/v2';

export type Game = {
  game_id: number;
  game_title: string;
  genre: string[];
  genre_id: number;
  platform: string[];
  game_description: string;
};

export type Review = {
  review_id: number;
  game_id: string;
  external_game_id: string[];
  game_title: number;
  genre: string[];
  platform: string;
  review_title: string;
  text: string;
  user_id: number;
  rating: number;
  published: boolean;
};

class GameService {
  game: Game = {
    game_id: 0,
    game_title: '',
    genre: [],
    genre_id: 0,
    platform: [],
    game_description: '',
  };
  games: Game[] = [];

  /**
   * Get game with given id.
   */
  get(id: number) {
    return axios.get<Game>('/games/' + id).then((response) => response.data);
  }

  /**
   * Get all games.
   */
  getAll() {
    return axios.get<Game[]>('/games').then((response) => response.data);
  }

  /**
   * Create new task having the given title.
   *
   * Resolves the newly created task id.
   */
  create(title: string, genre: string, platform: number, description: string) {
    return axios
      .post<{ id: number }>('/games', {
        game_title: title,
        genre: genre,
        platform: platform,
        game_description: description,
      })
      .then((response) => response.data.id);
  }
}

class ReviewService {
  review: Review = {
    review_id: 0,
    game_id: '',
    external_game_id: [],
    game_title: 0,
    genre: [],
    platform: '',
    review_title: '',
    text: '',
    user_id: 0,
    rating: 0,
    published: false,
  };
  reviews: Review[] = [];

  /**
   * Get review with given id.
   */
  get(id: number) {
    return axios.get<Review>('/reviews/' + id).then((response) => response.data);
  }

  /**
   * Get all reviews.
   */
  getAll() {
    return axios.get<Review[]>('/reviews').then((response) => response.data);
  }

  /**
   * Create new review having the given title.
   *
   * Resolves the newly created review id.
   */
  create(title: string, text: string, rating: number) {
    return axios
      .post<{ id: number }>('/reviews', { review_title: title, text: text, rating: rating })
      .then((response) => response.data.id);
  }
}

export let gameService = sharedComponentData(new GameService());
export let reviewService = sharedComponentData(new ReviewService());
