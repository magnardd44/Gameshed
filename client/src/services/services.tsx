import axios from 'axios';
import { sharedComponentData } from 'react-simplified';

axios.defaults.baseURL = 'http://localhost:3000/api/v2';

export type Game = {
  game_id: number;
  game_title: string;
  genre: string[];
  genre_id: number;
  platform: string[];
  game_description: string;
};

class GameService {
  game: Game = {
    game_id: 0,
    game_title: '',
    genre: [],
    genre_id: 0,
    platform: [],
    game_description: '',
  };
  games: Game[] = [];

  /**
   * Get game with given id.
   */
  get(id: number) {
    return axios.get<Game>('/games/' + id).then((response) => response.data);
  }

  /**
   * Get all games.
   */
  getAll() {
    return axios.get<Game[]>('/games').then((response) => response.data);
  }

  /**
   * Create new task having the given title.
   *
   * Resolves the newly created task id.
   */
  create(title: string, genre: string, platform: number, description: string) {
    return axios
      .post<{ id: number }>('/games', {
        game_title: title,
        genre: genre,
        platform: platform,
        game_description: description,
      })
      .then((response) => response.data.id);
  }
}

export let gameService = sharedComponentData(new GameService());
