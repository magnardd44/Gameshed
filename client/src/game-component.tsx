import * as React from 'react';
import { Component } from 'react-simplified';
import {
  Alert,
  Card,
  Row,
  Column,
  Form,
  Button,
  ThumbNail,
  Container,
  ColumnCentre,
  Linebreak,
} from './widgets';
import { NavLink } from 'react-router-dom';
import { gameService, reviewService, Game, Review } from './services';
import { createHashHistory } from 'history';

const history = createHashHistory(); // Use history.push(...) to programmatically change path, for instance after successfully saving a student

export class GameCard extends Component {
  game: Game = {
    game_id: 0,
    game_title: '',
    genre: [],
    genre_id: 0,
    platform: [],
    game_description: '',
  };
  description =
    'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco...';
  render() {
    return (
      <Card title={this.game.game_title}>
        <h6 className="card-subtitle mb-2 text-muted">
          Terningkast:
          <ThumbNail small img="https://cdn-icons-png.flaticon.com/512/220/220725.png"></ThumbNail>
        </h6>
        <Row>
          <Column width={2}>
            <ThumbNail img="https://cdn-icons-png.flaticon.com/512/686/686589.png"></ThumbNail>
          </Column>
          <Column width={6}>
            {this.game.game_description}
            <Linebreak></Linebreak>
          </Column>
          <Column width={2}></Column>
        </Row>
        <Linebreak></Linebreak>
        <Row>
          <Column>Sjanger: {this.game.genre}</Column>
        </Row>
        <Linebreak></Linebreak>
        <Row>
          <Column>Platform: {this.game.platform}</Column>
        </Row>
      </Card>
    );
  }
  /** return axios
   */
  mounted() {
    gameService.getAll().then((result) => {
      this.game = result[0];
      console.log(this.game);
    });
  }
}
