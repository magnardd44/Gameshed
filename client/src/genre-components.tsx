import * as React from 'react';
import { Component, sharedComponentData } from 'react-simplified';
import {
  Alert,
  Card,
  Row,
  Column,
  Form,
  Button,
  ThumbNail,
  Container,
  ColumnCentre,
  Linebreak,
  FormContainer,
  FormGroup,
  CategoryCard,
} from './widgets';
import { NavLink } from 'react-router-dom';
import { gameService, Game } from './services/game-services';
import { Genre } from './services/genre-service';
import { createHashHistory } from 'history';
import axios from 'axios';
import { genreService } from './services/genre-service';
import { Review, reviewService } from './services/review-service';

const history = createHashHistory();

export class Category extends Component {
  genres: Genre[] = [];
  reviews: Review[] = [];
  games: Game[] = [];
  game: Game = {
    game_id: 0,
    game_title: '',
    genre: 0,
    genre_id: 0,
    platform: 0,
    game_description: '',
    igdb_id: 0,
    genres: [],
    platforms: [],
  };
  review: Review = {
    review_id: 0,
    game_id: 0,
    game_title: '',
    review_title: '',
    text: '',
    user_id: 0,
    rating: 0,
    published: false,
    genre_id: 0,
    relevant: 0,
    platform_id: 0,
  };
  state = { isHidden: false };
  genreCall(id: number) {
    reviewService
      .getGenre(id)
      .then((data) => {
        console.log(data);
        this.reviews = data;
      })
      .catch((error) => Alert.danger('Error retrieving reviews: ' + error.message));
    this.setState({ isHidden: false });
  }
  render() {
    return (
      <Container>
        Velg kategori/sjanger:
        <Row>
          <ColumnCentre width={12} smwidth={4} mdwidth={2}>
            {this.genres.map((genre) => (
              <CategoryCard
                key={genre.genre_id}
                img={
                  genre.genre_img
                    ? 'https://helenaagustsson.github.io/INFT2002-images/images/' +
                      genre.genre_img +
                      '.png'
                    : 'https://helenaagustsson.github.io/INFT2002-images/images/' +
                      genre.genre_name +
                      '.png'
                }
              >
                <Button.Success onClick={() => this.genreCall(genre.genre_id)}>
                  {genre.genre_name}
                </Button.Success>
              </CategoryCard>
            ))}
          </ColumnCentre>
          <ColumnCentre>
            {this.state.isHidden ? null : (
              <div className="sticky-top">
                {this.reviews.map((review, index) => (
                  <Row key={index}>
                    <Column>{review.game_title}</Column>
                    <Column>
                      <NavLink to={'/publishedReviews/' + review.review_id}>
                        {review.review_title}
                      </NavLink>
                    </Column>
                    <Column>{review.rating}</Column>
                  </Row>
                ))}
              </div>
            )}
          </ColumnCentre>
        </Row>
      </Container>
    );
  }
  mounted() {
    genreService
      .getAll()
      .then((genres) => {
        this.genres = genres;
      })
      .catch((error) => Alert.danger('Error getting reviews: ' + error.message));
  }
}
