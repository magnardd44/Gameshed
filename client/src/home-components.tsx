import * as React from 'react';
import { Component, sharedComponentData } from 'react-simplified';
import {
  Alert,
  Card,
  Row,
  Column,
  Form,
  Button,
  ThumbNail,
  Container,
  ColumnCentre,
  Linebreak,
  FormContainer,
  FormGroup,
  CategoryCard,
} from './widgets';
import { NavLink } from 'react-router-dom';
import { Genre } from './services/genre-service';
import { createHashHistory } from 'history';
import axios from 'axios';
import { genreService } from './services/genre-service';
import userService from './services/user-service';

export const history = createHashHistory();

export class Home extends Component {
  render() {
    return (
      <Container textalign={'center'}>
        <Row>
          <ColumnCentre width={6}>
            <Card title={'Søk etter spill'}>
              <Button.Success
                onClick={() => {
                  history.push('/search');
                }}
              >
                Søk
              </Button.Success>
            </Card>
          </ColumnCentre>
          <ColumnCentre width={6}>
            <Card title={'Søk etter anmeldelser'}>
              <Button.Success
                onClick={() => {
                  history.push('/reviews');
                }}
              >
                Søk
              </Button.Success>
            </Card>
          </ColumnCentre>
        </Row>
        <Linebreak></Linebreak>
        <Row>
          <ColumnCentre width={userService.token ? 12 : 6}>
            <Card title={'Legg til nytt spill'}>
              <Button.Success
                onClick={() => {
                  history.push('/addGame');
                }}
              >
                Legg til spill
              </Button.Success>
            </Card>
          </ColumnCentre>
          {userService.token ? (
            ''
          ) : (
            <ColumnCentre width={6}>
              <Card title="Registrer ny konto">
                <Button.Success
                  onClick={() => {
                    history.push('/user');
                  }}
                >
                  Registrer deg
                </Button.Success>
              </Card>
            </ColumnCentre>
          )}
        </Row>
      </Container>
    );
  }
}

export class ReviewHome extends Component {
  render() {
    return (
      <Container textalign={'center'}>
        <Row>
          <ColumnCentre width={6}>
            <Card title={'Sorter etter sjanger'}>
              <Button.Success
                onClick={() => {
                  history.push('/reviews-by-genre');
                }}
              >
                Søk
              </Button.Success>
            </Card>
          </ColumnCentre>
          <ColumnCentre width={6}>
            <Card title={'Sorter etter plattform'}>
              <Button.Success
                onClick={() => {
                  history.push('/reviews-by-platform');
                }}
              >
                Søk
              </Button.Success>
            </Card>
          </ColumnCentre>
        </Row>
        <Linebreak></Linebreak>
        <Row>
          <ColumnCentre width={6}>
            <Card title={'Topp ti anmeldelser'}>
              <Button.Success
                onClick={() => {
                  history.push('/reviews-by-stars');
                }}
              >
                Søk
              </Button.Success>
            </Card>
          </ColumnCentre>
          <ColumnCentre width={6}>
            <Card title={'Siste ti anmeldelser?'}>
              <Button.Success
                onClick={() => {
                  history.push('/reviews-by-date');
                }}
              >
                Søk
              </Button.Success>
            </Card>
          </ColumnCentre>
        </Row>
        <Linebreak></Linebreak>
        <Row>
          <ColumnCentre width={12}>
            <Card title={'Alle anmeldelser'}>
              <Button.Success
                onClick={() => {
                  history.push('/publishedReviews');
                }}
              >
                Alle anmeldelser
              </Button.Success>
            </Card>
          </ColumnCentre>
        </Row>
      </Container>
    );
  }
}
