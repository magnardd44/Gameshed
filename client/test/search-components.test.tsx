import * as React from 'react';
import ReactDOM from 'react-dom';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

import { shallow } from 'enzyme';
import {
  history,
  Search,
  SearchListings,
  SearchResult,
  SearchHotBar,
} from '../src/search-components';
import { gameService } from '../src/services/game-service';
import { ColumnCentre, Form, Card, Alert, Button, Container, Column } from '../src/widgets';

jest.setTimeout(1000);

const mockAdapter = new MockAdapter(axios);

Alert.info = jest.fn(() => {});
//Alert.warning = jest.fn(() => {});
//Alert.success = jest.fn(() => {});

const testGameComplete = {
  game_id: 1,
  igdb_id: 22,
  game_title: 'Title',
  genre: ['Adventure', 'Shooter'],
  platform: ['PC', 'Playstation', 'xbox'],
  game_description: 'Text',
  igdb: {
    cover_url: 'url',
    aggregated_rating: 50,
    screenshots_url: ['screen1', 'screen2'],
    similar_games: [{ id: 1, name: 'similar' }],
    release_date: 2000,
  },
};

const testGameEmpty = {
  game_id: 0,
  igdb_id: 0,
  game_title: '',
  genre: [],
  platform: [],
  game_description: '',
  igdb: null,
};

const testGameNone = {};

const testGames = [{ game_id: 1, igdb_id: 10, game_title: 'Zelda' }];
const testGames2 = [{ game_id: 2, igdb_id: 22, game_title: 'Mario' }];

describe('SearchHotBar component', () => {
  test('Search component default', () => {
    //@ts-ignore
    const wrapper = shallow(<SearchHotBar key="1" />);
    expect(wrapper).toHaveProperty('key');
    expect(wrapper).toMatchSnapshot();
  });
});

describe('Search component', () => {
  beforeEach(() => {
    jest.useRealTimers();
  });

  test('Search component default', () => {
    const wrapper = shallow(<Search />);

    expect(wrapper).toMatchSnapshot();

    //@ts-ignore
    expect(wrapper.instance().input).toEqual('');
    //@ts-ignore
    expect(wrapper.instance().lastInput).toEqual('');
  });

  test('Search db for Zelda', (done) => {
    const wrapper = shallow(<Search />);

    const spy = jest.spyOn(gameService, 'search_db');
    spy.mockImplementationOnce(() => {});

    wrapper.find(Form.Input).simulate('change', { currentTarget: { value: 'Zelda' } });

    setTimeout(() => {
      expect(gameService.search_db).toBeCalled();
      spy.mockRestore();
      done();
    });
  });

  test('IGDB searcher is running', (done) => {
    jest.useFakeTimers();
    const wrapper = shallow(<Search />);

    const spyDB = jest.spyOn(gameService, 'search_db');
    spyDB.mockImplementationOnce(() => {});

    const spyIGDB = jest.spyOn(gameService, 'search_igdb');
    spyIGDB.mockImplementation(() => {});

    wrapper.find(Form.Input).simulate('change', { currentTarget: { value: 'Mario' } });

    expect(gameService.search_igdb).not.toBeCalled();
    jest.advanceTimersByTime(500);
    expect(gameService.search_igdb).toBeCalled();

    spyDB.mockRestore();
    spyIGDB.mockRestore();

    done();
  });

  test('IGDB searcher is stopped on unmount', (done) => {
    jest.useFakeTimers();
    const wrapper = shallow(<Search />);

    const spyDB = jest.spyOn(gameService, 'search_db');
    spyDB.mockImplementation(() => {});

    const spyIGDB = jest.spyOn(gameService, 'search_igdb');
    spyIGDB.mockImplementation(() => {});

    wrapper.find(Form.Input).simulate('change', { currentTarget: { value: 'Mario' } });

    expect(gameService.search_igdb).not.toBeCalled();
    jest.advanceTimersByTime(500);
    expect(gameService.search_igdb).toHaveBeenCalledTimes(1);
    wrapper.find(Form.Input).simulate('change', { currentTarget: { value: 'Zelda' } });
    wrapper.unmount();
    jest.advanceTimersByTime(500);
    expect(gameService.search_igdb).toHaveBeenCalledTimes(1);

    spyDB.mockRestore();
    spyIGDB.mockRestore();

    done();
  });

  test('No IGDB searcher to be stopped on unmount', (done) => {
    const wrapper = shallow(<Search />);

    //@ts-ignore
    clearInterval(wrapper.instance().igdbSearcher);
    //@ts-ignore
    wrapper.instance().igdbSearcher = null;

    const spy = jest.spyOn(window, 'clearInterval');

    wrapper.unmount();

    setTimeout(() => {
      expect(clearInterval).not.toBeCalled();
      spy.mockRestore();

      done();
    });
  });

  test('IGDB results is cleared on empty input', (done) => {
    jest.useFakeTimers();
    const wrapper = shallow(<Search />);

    const spyDB = jest.spyOn(gameService, 'search_db');
    spyDB.mockImplementation(() => {});

    const spyIGDB = jest.spyOn(gameService, 'search_igdb');
    spyIGDB.mockImplementation(() => {});

    wrapper.find(Form.Input).simulate('change', { currentTarget: { value: 'IGDB empty' } });

    expect(gameService.search_igdb).not.toBeCalled();
    jest.advanceTimersByTime(500);
    expect(gameService.search_igdb).toHaveBeenCalledTimes(1);

    wrapper.find(Form.Input).simulate('change', { currentTarget: { value: 'Zelda' } });
    jest.advanceTimersByTime(500);
    expect(gameService.search_igdb).toHaveBeenCalledTimes(2);

    wrapper.find(Form.Input).simulate('change', { currentTarget: { value: '' } });
    jest.advanceTimersByTime(500);
    expect(gameService.search_igdb).toHaveBeenCalledTimes(2);

    spyDB.mockRestore();
    spyIGDB.mockRestore();

    done();
  });

  test('Press enter on input field', (done) => {
    const wrapper = shallow(<Search />);

    wrapper.find(Form.Input).simulate('keyup', { key: 'Space' });
    wrapper.find(Form.Input).simulate('keyup', { key: 'Enter' });

    setTimeout(() => {
      expect(Alert.info).toHaveBeenCalledTimes(1);
      done();
    });
  });

  test('Press search button with input', (done) => {
    const wrapper = shallow(<Search />);

    const spy = jest.spyOn(history, 'push');
    mockAdapter.onGet('games/search/Zelda').reply(200, []);
    mockAdapter.onPost('search').reply(200, []);
    mockAdapter.onPost('search/get_igdb_extra').reply(200, {});

    //@ts-ignore
    wrapper.instance().input = 'Zelda';
    wrapper.find(Button.Success).simulate('click');

    setTimeout(() => {
      expect(history.push).toBeCalled();
      spy.mockRestore();
      done();
    });
  });

  test('Press search button with empty input', (done) => {
    const wrapper = shallow(<Search />);
    wrapper.find(Button.Success).simulate('click');

    setTimeout(() => {
      expect(Alert.info).toBeCalled();
      done();
    });
  });

  test('Search result is displayed', (done) => {
    const wrapper = shallow(<Search />);

    const searchGame = 'ZeldaMario';

    mockAdapter.onGet('games/search/' + searchGame).reply(200, testGames);
    mockAdapter.onPost('search').reply(200, testGames2);

    let elementsToMatch = [
      //@ts-ignore
      <SearchHotBar>Zelda</SearchHotBar>,
      //@ts-ignore
      <SearchHotBar>Mario</SearchHotBar>,
    ];

    wrapper.find(Form.Input).simulate('change', { currentTarget: { value: searchGame } });

    setTimeout(() => {
      expect(wrapper.containsAllMatchingElements(elementsToMatch)).toEqual(true);
      done();
    }, 350);
  });

  test('Search result is clicked', (done) => {
    const wrapper = shallow(<Search />);

    const spy = jest.spyOn(history, 'push');
    mockAdapter.onGet('games/' + testGames[0].game_id).reply(200, testGames);
    mockAdapter.onGet('games/search/' + testGames[0].game_title).reply(200, testGames);
    mockAdapter.onPost('search').reply(200, testGames2);

    wrapper.find(Form.Input).simulate('change', { currentTarget: { value: 'Zelda' } });

    setTimeout(() => {
      //@ts-ignore
      expect(wrapper.containsMatchingElement(<SearchHotBar>Zelda</SearchHotBar>)).toBe(true);
      wrapper.find(SearchHotBar).simulate('click');

      expect(history.push).toBeCalled();
      spy.mockRestore();
      done();
    });
  });
});

describe('SearchListings component', () => {
  test('SearchListings default', (done) => {
    mockAdapter.onGet('/genres').reply(200, ['genre1', 'genre2']);
    mockAdapter.onGet('/platforms').reply(200, ['platform1', 'platform2']);

    gameService.db = [];
    gameService.igdb = [];

    //@ts-ignore
    const wrapper = shallow(<SearchListings />);

    setTimeout(() => {
      expect(wrapper).toMatchSnapshot();
      done();
    });
  });

  test('Test select filters', (done) => {
    mockAdapter.onGet('/genres').reply(200, ['genre1', 'genre2']);
    mockAdapter.onGet('/platforms').reply(200, ['platform1', 'platform2']);

    //@ts-ignore
    gameService.db = [testGameComplete, testGameComplete];
    //@ts-ignore
    gameService.igdb = [testGameComplete, testGameComplete, testGameEmpty];

    //@ts-ignore
    const wrapper = shallow(<SearchListings />);

    expect(wrapper.find(SearchResult)).toHaveLength(5);

    done();
  });

  test('Test select filters', (done) => {
    mockAdapter.onGet('/genres').reply(200, ['genre1', 'genre2']);
    mockAdapter.onGet('/platforms').reply(200, ['platform1', 'platform2']);

    //@ts-ignore
    const wrapper = shallow(<SearchListings />);

    expect(wrapper.find(Form.Select)).toHaveLength(3);

    wrapper
      .find(Form.Select)
      .at(0)
      .simulate('change', { currentTarget: { value: 'genre2' } });
    wrapper
      .find(Form.Select)
      .at(1)
      .simulate('change', { currentTarget: { value: 'platform2' } });
    wrapper
      .find(Form.Select)
      .at(2)
      .simulate('change', { currentTarget: { value: 1900 } });

    //@ts-ignore
    expect(wrapper.instance().genre).toEqual('genre2');
    //@ts-ignore
    expect(wrapper.instance().platform).toEqual('platform2');
    //@ts-ignore
    expect(wrapper.instance().year).toEqual(1900);

    wrapper
      .find(Form.Select)
      .at(0)
      .simulate('change', { currentTarget: { value: 'alle' } });
    wrapper
      .find(Form.Select)
      .at(1)
      .simulate('change', { currentTarget: { value: 'alle' } });
    wrapper
      .find(Form.Select)
      .at(2)
      .simulate('change', { currentTarget: { value: 'alle' } });

    //@ts-ignore
    expect(wrapper.instance().genre).toEqual('alle');
    //@ts-ignore
    expect(wrapper.instance().platform).toEqual('alle');
    //@ts-ignore
    expect(wrapper.instance().year).toEqual(0);

    done();
  });

  test('Test addGame button', (done) => {
    const wrapper = shallow(<SearchListings />);

    mockAdapter.onGet('/genres').reply(200, ['genre1', 'genre2']);
    mockAdapter.onGet('/platforms').reply(200, ['platform1', 'platform2']);

    const spy = jest.spyOn(history, 'push');

    wrapper.find(Button.Success).simulate('click');

    setTimeout(() => {
      expect(history.push).toBeCalled();
      spy.mockRestore();
      done();
    });
  });

  test('Filter when games are incomplete', (done) => {
    const wrapper = shallow(<SearchListings />);

    //@ts-ignore
    gameService.db = [testGameNone];
    //@ts-ignore
    gameService.igdb = [testGameNone];

    setTimeout(() => {
      //Three Form.Select with only the option 'alle' each
      expect(wrapper.find('option')).toHaveLength(3);
      done();
    });
  });

  test('Filter is sorting on year', (done) => {
    const wrapper = shallow(<SearchListings />);

    gameService.db = [
      //@ts-ignore
      { igdb: { release_date: new Date('1980').valueOf() / 1000 } }, //@ts-ignore
      { igdb: { release_date: new Date('2010').valueOf() / 1000 } }, //@ts-ignore
      { igdb: { release_date: new Date('1990').valueOf() / 1000 } },
    ];

    gameService.igdb = [
      //@ts-ignore
      { igdb: { release_date: new Date('2020').valueOf() / 1000 } }, //@ts-ignore
      { igdb: { release_date: new Date('1960').valueOf() / 1000 } }, //@ts-ignore
      { igdb: { release_date: new Date('1990').valueOf() / 1000 } },
    ];

    setTimeout(() => {
      expect(wrapper.find('option')).toHaveLength(8);
      expect(
        wrapper
          .find('option')
          .at(0)
          .matchesElement(<option>alle</option>)
      ).toBe(true);
      expect(
        wrapper
          .find('option')
          .at(1)
          .matchesElement(<option>alle</option>)
      ).toBe(true);
      expect(
        wrapper
          .find('option')
          .at(2)
          .matchesElement(<option>alle</option>)
      ).toBe(true);
      expect(
        wrapper
          .find('option')
          .at(3)
          .matchesElement(<option>1960</option>)
      ).toBe(true);
      expect(
        wrapper
          .find('option')
          .at(4)
          .matchesElement(<option>1980</option>)
      ).toBe(true);
      expect(
        wrapper
          .find('option')
          .at(5)
          .matchesElement(<option>1990</option>)
      ).toBe(true);
      expect(
        wrapper
          .find('option')
          .at(6)
          .matchesElement(<option>2010</option>)
      ).toBe(true);
      expect(
        wrapper
          .find('option')
          .at(7)
          .matchesElement(<option>2020</option>)
      ).toBe(true);

      done();
    });
  });
});

describe('SearchResult component', () => {
  test('SearchResult default', () => {
    //@ts-ignore
    const wrapper = shallow(<SearchResult game={testGameComplete} />);

    expect(wrapper).toMatchSnapshot();
  });

  test('SearchResult with empty game', () => {
    const wrapper = shallow(<SearchResult game={testGameEmpty} />);

    expect(wrapper).toMatchSnapshot();
  });

  test('Click on button', (done) => {
    //@ts-ignore
    const wrapper = shallow(<SearchResult game={testGameComplete} />);
    const spy = jest.spyOn(history, 'push');

    wrapper.find(Button.Success).simulate('click');

    setTimeout(() => {
      expect(history.push).toBeCalled();
      spy.mockRestore();
      done();
    });
  });
});
