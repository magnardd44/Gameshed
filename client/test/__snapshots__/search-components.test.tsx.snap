// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component Search component default 1`] = `
<Fragment>
  <Container
    textalign="centre"
  >
    <Row>
      <ColumnCentre
        offset={2}
        width={6}
      >
        <FormInput
          onChange={[Function]}
          onKeyUp={[Function]}
          placeholder="Søk etter et spill"
          type=""
          value=""
        />
      </ColumnCentre>
      <ColumnCentre
        width={4}
      >
        <ButtonSuccess
          onClick={[Function]}
        >
          Søk
        </ButtonSuccess>
      </ColumnCentre>
    </Row>
  </Container>
</Fragment>
`;

exports[`SearchHotBar component Search component default 1`] = `
<div
  className=" option"
  role="option"
  style={
    Object {
      "backgroundColor": "lightgray",
      "border": "1px solid black",
      "borderRadius": "5px",
      "cursor": "pointer",
      "marginTop": "10px",
    }
  }
/>
`;

exports[`SearchListings component SearchListings default 1`] = `
<Fragment>
  <Container>
    Filtrer dine søkeresultater:
    <FormContainer>
      <FormGroup>
        <FormLabel>
          Sjanger: 
        </FormLabel>
        <FormSelect
          onChange={[Function]}
          value=""
        >
          <option
            key="0"
          />
          <option
            key="1"
          />
        </FormSelect>
        <FormLabel>
          Platform: 
        </FormLabel>
        <FormSelect
          onChange={[Function]}
          value=""
        >
          <option
            key="0"
          />
          <option
            key="1"
          />
        </FormSelect>
        <FormLabel>
          År: 
        </FormLabel>
        <FormSelect
          onChange={[Function]}
          value={0}
        >
          <option
            value="1"
          >
            2020
          </option>
        </FormSelect>
      </FormGroup>
    </FormContainer>
  </Container>
  <Container />
  <Container>
    Fant du ikke det du lette etter?
    <Linebreak />
    <ButtonSuccess
      onClick={[Function]}
    >
      Legg til nytt spill
    </ButtonSuccess>
  </Container>
</Fragment>
`;

exports[`SearchResult component SearchResult default 1`] = `
<Card
  title="Title"
>
  <h6
    className="card-subtitle mb-2 text-muted"
  >
    Terningkast:
    <ThumbNail
      img="https://cdn-icons-png.flaticon.com/512/220/220725.png"
      small={true}
    />
  </h6>
  <Row>
    <Column
      width={2}
    >
      <ThumbNail
        img="url"
      />
    </Column>
    <Column
      width={8}
    >
      Text
      <Linebreak />
    </Column>
    <Column
      width={2}
    >
       
      <ButtonSuccess
        onClick={[Function]}
      >
        Les mer
      </ButtonSuccess>
    </Column>
  </Row>
  <Row>
    Sjanger: 
    Adventure, Shooter
  </Row>
  <Row>
    Platformer: 
    PC, Playstation, xbox
  </Row>
  <Row>
    Utgivelsesår:
     
    1970
     
  </Row>
  <Linebreak />
</Card>
`;

exports[`SearchResult component SearchResult with empty game 1`] = `
<Card
  title=""
>
  <h6
    className="card-subtitle mb-2 text-muted"
  >
    Terningkast:
    <ThumbNail
      img="https://cdn-icons-png.flaticon.com/512/220/220725.png"
      small={true}
    />
  </h6>
  <Row>
    <Column
      width={2}
    >
      <ThumbNail
        img=""
      />
    </Column>
    <Column
      width={8}
    >
      Ingen beskrivelse
      <Linebreak />
    </Column>
    <Column
      width={2}
    >
       
      <ButtonSuccess
        onClick={[Function]}
      >
        Les mer
      </ButtonSuccess>
    </Column>
  </Row>
  <Row>
    Sjanger: 
  </Row>
  <Row>
    Platformer: 
  </Row>
  <Row>
    Utgivelsesår:
     
     
  </Row>
  <Linebreak />
</Card>
`;
